presto-main/src/main/java/io/prestosql/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -126,695 +122,988 @@ public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, Sym
+            PlanAndMappings rewrittenSubquery = node.getSubquery().accept(this, new UnaliasContext(mappingForSubquery));
+
+            // unify mappings from Input and Subquery to rewrite Subquery assignments
+            Map<Symbol, Symbol> resultMapping = new HashMap<>();
+            resultMapping.putAll(rewrittenInput.getSymbolMappings());
+            resultMapping.putAll(rewrittenSubquery.getSymbolMappings());
+            mapper = new SymbolMapper(resultMapping);
+
+            ImmutableList.Builder<Map.Entry<Symbol, Expression>> builder = ImmutableList.builder();
+            for (Map.Entry<Symbol, Expression> assignment : node.getSubqueryAssignments().entrySet()) {
+                builder.add(new SimpleEntry<>(mapper.map(assignment.getKey()), mapper.map(assignment.getValue())));
+            }
+            List<Map.Entry<Symbol, Expression>> rewrittenAssignments = builder.build();
 
-            return new SortNode(node.getId(), source, canonicalizeAndDistinct(node.getOrderingScheme()), node.isPartial());
+            // deduplicate assignments
+            Map<Symbol, Expression> deduplicateAssignments = new HashMap<>();
+            for (Map.Entry<Symbol, Expression> assignment : rewrittenAssignments) {
+                Expression previous = deduplicateAssignments.put(assignment.getKey(), assignment.getValue());
+                checkState(previous == null || assignment.getValue().equals(previous), "different expressions assigned to the same symbol");
+            }
+
+            // derive new mappings for Subquery assignments outputs
+            Map<Expression, Symbol> inputsToOutputs = new HashMap<>();
+            Map<Symbol, Symbol> newMapping = new HashMap<>();
+            for (Map.Entry<Symbol, Expression> assignment : deduplicateAssignments.entrySet()) {
+                Expression expression = assignment.getValue();
+                // 1. for trivial symbol projection, map output symbol to input symbol
+                if (expression instanceof SymbolReference) {
+                    Symbol value = Symbol.from(expression);
+                    if (!assignment.getKey().equals(value)) {
+                        newMapping.put(assignment.getKey(), value);
+                    }
+                }
+                // 2. map same deterministic expressions within a projection into the same symbol
+                // omit NullLiterals since those have ambiguous types
+                else if (DeterminismEvaluator.isDeterministic(expression, metadata) && !(expression instanceof NullLiteral)) {
+                    Symbol previous = inputsToOutputs.get(expression);
+                    if (previous == null || assignment.getKey().equals(previous)) {
+                        inputsToOutputs.put(expression, assignment.getKey());
+                    }
+                    else {
+                        newMapping.put(assignment.getKey(), previous);
+                    }
+                }
+            }
```

Comment:
This is duplicated in a couple of places. You could extract it to a separate method.